### PR TITLE
[models,trainers,hparams] refactor: infrastructure layer for multi-component models and multi-variant algorithms

### PR DESCRIPTION
**Pull request by Jayce-Ping**
## Summary

Infrastructure layer for multi-component models and multi-variant algorithms. Model code owns caller-named component variants and snapshots; trainers own role meaning, scheduling, optimization, and checkpoint compatibility.

## What lands

- Structured trajectory/state interfaces and ordered heterogeneous references.
- Component variants, named snapshots, one prepared `ModelBundle`, and role-aware checkpoint manifests.
- Per-role optimizer configuration, AdamW/Muon loading, and startup backend validation.
- Shared trainer primitives for forward kwargs, state validation, replay batching, H2D prefetch, coupled replay, and decoupled timestep loops.
- Multi-role backend/checkpoint responsibilities moved out of `BaseTrainer`.
- Packed eval/CRD/statistics collectives and benchmark-gated `gather_samples` packing.
- Explicit memory-placement policy: prepared trainable variants remain backend-owned; component, EMA/snapshot, and sample offload paths are documented.
- `frozen_parameters_dtype` accepts a scalar or default/group/component mapping, allowing checkpoint-native VAE precision without changing transformer or text-encoder policy.
- Flow timestep-to-sigma mapping now uses a float64 intermediate so strict float32 boundary interiors cannot round back to sigma one on CUDA.

## Compatibility and performance

- Existing algorithm hooks and batch-key precedence remain stable.
- Critical guards remain intact: FSDP2 identity rebind, ZeRO role isolation, PEFT trainability restoration, checkpoint manifest compatibility, replay diagnostics, and rank-uniform metric keys.
- Matched 16-GPU NFT A/B, excluding warm-up: median training step **0.7492s → 0.7069s** (5.6% faster).
- Peak memory on the same config: **12627 MiB → 12625 MiB**.
- Packed collective A/B kept only winning paths; mixed-dtype CRD and large CPU-bound payloads fall back to their faster field-wise path.

## Verification

- [x] Infra suite: **1005 passed**.
- [x] 16-GPU SD3.5 LoRA sanity passes under both DDP and FSDP2 with `transformers/text_encoders=bf16`, `vae=fp32`, and exact on-policy replay.
- [x] 16-GPU, 2-node GRPO and NFT correctness/performance runs.
- [x] Full child matrix on #210: **40/40** combinations passed real backward/optimizer step plus model-only checkpoint save and resume.
- [x] MiniMax child suite after restacking: **1558 passed, 14 skipped**.
- [x] Final publishability review sanitized multi-node and HDFS defaults; **91 targeted tests passed**.

## Stack

Merge order: **#209 → #210 → #212**.

Current head: `98b6b4d`.

### CODE REVIEW

**Comment by Jayce-Ping** [comment]
## Verification complete — ready to merge

Everything in the test plan above now has a result. Six defects surfaced only under real
multi-GPU execution and are fixed on this branch; unit tests and config parsing had passed
through all of them.

| commit | defect |
|---|---|
| `87755c7` | LTX2 golden compared with `atol=0, rtol=0`, so it reported the machine, not the code. Drift measured at 1.192e-07 — exactly one fp32 ULP. Present since the test landed in `310be3b`, not a later regression. |
| `44c70bb` | Every shipped `optimizers:` entry is named `default` while the role is `base`; the single-entry fallback returned it unrenamed and `build_optimizer` found no parameters. **All 54 such configs died at startup**, on both DeepSpeed and FSDP. |
| `0a3aa32` | Muon was allowed on FSDP1, which flattens to a 1D FlatParameter and fails at the first step — after a full rollout is already paid for. |
| `68d9937` | `clip_grad_norm_` received only the trainable subset, so accelerate never delegated to FSDP's collective clip and each rank clipped by its own shard norm. FSDP1 grad_norm 0 -> 6e-4, now identical to ZeRO-2/DDP/FSDP2. |
| `a23663a` | LoRA checkpoint saving called `FSDP.state_dict_type` on a component, i.e. a non-root FSDP instance, poisoning `_is_root` and killing the next rollout an epoch later. Gathering goes through the prepared root now. |
| `bbdfe48` | Added a trainability guard after `prepare`; `68d9937` corrected it to be collective. |

One correction worth recording: `Trainable: 0` under FSDP is **not** a failure. FSDP splits a
flattened unit by byte range, so with LoRA the frozen base fills the early shards and the whole
adapter can land on one rank — an 8-rank probe put all 2048 trainable elements on rank 7 and none
on ranks 0-6. The log line is now labelled as this rank's shard. This behavior predates the branch;
verified identical at the merge-base.

Remaining known limitations, none blocking:

- FSDP1 + Muon is refused by design; use FSDP2 or DDP.
- Multi-variant / multi-role training is untested because no shipped algorithm owns two roles.
  It arrives with DMD/TDM.
- The four Muon tests require torch >= 2.9.

**Comment by Jayce-Ping** [comment]
## Checkpoint layout, and the resume it was quietly breaking

Three commits since the last update, all in the same area.

A checkpoint reconstructed its directory layout in four independent places --
`save_checkpoint`, `_load_lora`, `_load_full_model` and `_detect_checkpoint_type` -- so
nothing kept the writer and the readers agreeing. They now share one
`_checkpoint_entries` resolver and the result is recorded in a `manifest.json`, which is
authoritative when present and reconstructed from the legacy rule when it is not, so
existing checkpoints keep loading unchanged.

Multi-role algorithms were paying for the drift. Only the base variant was ever written,
so a run that saved every five epochs and later resumed came back with its generator
restored beside a freshly initialized fake score, trained against the wrong critic and
reported nothing wrong. Two further ordering bugs sat behind it: the weight-only resume
runs while the adapter is built, before the trainer declares roles, so both roles were
being loaded into the single live adapter and the second overwrote the first; and
`_mix_precision` never reached the late-declared variants, leaving a bf16 generator beside
an fp32 fake score.

Layout, unchanged for the single-component single-role case that every released checkpoint
uses:

```
checkpoint-N/
  manifest.json
  adapter_config.json          # primary artifact, PeftModel.from_pretrained reads it directly
  adapter_model.safetensors
  roles/fake/adapter_*         # training-only roles
  transformer_2/...            # multi-component, as before, plus its own roles/
```

## Verification

- [x] 17 round-trip tests: single/multi component x single/multi role x LoRA/full, each
      saving then loading and asserting the weights land on the right variant
- [x] Manifest-free checkpoints still load; a future `format_version` is refused
- [x] Full-weight loads stopped replacing the component object, which detached it from the
      prepared root and the registry
- [x] 953 tests pass (the four Muon ones need torch >= 2.9, as before)
- [x] 16-GPU Flux2-Klein-9B DMD2: checkpoint carries 240 generator tensors and 240 fake
      tensors, same-named tensors differ, and a resume loads `role=generator` then
      `role=fake` into their own variants
- [x] Fake LoRA is now bf16 on disk, not fp32

Mergeable: no conflicts with `main`, branch is not behind.
